torch/nn/modules/module.py

```diff
@@ -193,13 +195,68 @@ def _apply(self, fn):
+
+            # We subtract the refcount by 1 here, because the refcount includes the temporary reference
+            # `obj` as an argument to the `sys.getrefcount(obj)` function call.
+            # (For more details, see: https://docs.python.org/3/library/sys.html#sys.getrefcount)
+            param_extra_refcount = sys.getrefcount(self._parameters[key]) - 1
+
+            # We create `param` here after obtaining the refcount of `self._parameters[key]`, instead
+            # of using `for key, param in self._parameters.items():`, because the latter approach would
+            # unnecessarily bump the refcount of the `self._parameters[key]` tensor, which is undesirable.
+            param = self._parameters[key]
             if param is not None:
                 # Tensors stored in modules are graph leaves, and we don't
                 # want to create copy nodes, so we have to unpack the data.
-                param.data = fn(param.data)
-                if param._grad is not None:
-                    param._grad.data = fn(param._grad.data)
+                param_applied = fn(param)
+
+                # NOTE [ TensorImpl Type and Parameter Replacement ]
+                #
+                # If `param_applied` has the same TensorImpl type as `param`, we can use
+                # `param.data = param_applied` to change the value of `param` in place.
+                # Otherwise, we have to replace the tensor stored in `self._parameters[key]`
+                # with `param_applied`, which will break all previous references to
+                # the `self._parameters[key]` tensor. Similarly, depending on whether
+                # `grad_applied` has the same TensorImpl type as `param.grad`, we may be
+                # able to change the value of `param.grad` in place, or we may have to replace
+                # `param.grad` yf225 TODO finish this!
+                #
+                # In order to provide the same behavior for 
+                if param._is_same_impl_type(param_applied):
+                    if param.device != param_applied.device:
+                        if param_extra_refcount > 0:
+                            warnings.warn("Attempted to change device of `{}`, but there are \
+{} extra reference(s) to that tensor which would be broken by the device change, in future releases \
+of PyTorch.".format(key, param_extra_refcount))
+                    self._parameters[key].data = param_applied
+                else:
+                    # Tensors stored in modules are graph leaves, and we don't want to
+                    # create copy nodes, so we have to use `with torch.no_grad():` here.
+                    with torch.no_grad():
+                        # We use `.requires_grad_()` here to make sure the new `param` still
+                        # has the same `requires_grad` value as the old `param`.
+                        self._parameters[key] = param_applied.requires_grad_(param.requires_grad)
```

Comment:
I don't think this is correct, won't the parameter not even be an `nn.Parameter` anymore?